### PR TITLE
Support torch.onnx.dynamo_export within FakeTensorMode

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105477

Currently, exporting a model to ONNX with fake tensor mode requires the
user to load data and model within `torch.onnx.enable_fake_mode` context,
but the actual call to `torch.onnx.dynamo_export` is done outside such
context.

With this PR, we enable `torch.onnx.dynamo_export` to be called either
within `torch.onnx.enable_fake_mode` or outside of it. This feature
required changes to the core PyTorch Dynamo, which were greatly
supported by @ezyang

In future steps we will determine which scenario we are going to
support, but for now we can use either to explore different options and
scenarios and asses their pros and cons.

This PR also creates a separate suite of tests for fake mode specific
scenarios (`TestFxToOnnxFakeTensorWithOnnxRuntime`).
It was done separately to decrease the test time, but we
could merge it with the default `TestFxToOnnxWithOnnxRuntime`. The
additional parameters are `load_checkpoint_during_init` and
`export_within_fake_mode`

With the newly added supported of nested export within fake mode, the
following scenarios are now supported:

```python
import torch

with torch.onnx.enable_fake_mode() as fake_context:
    fake_args = create_args()
    fake_kwargs = create_kwargs()
    fake_model = create_model()
    fake_model.load_state_dict(torch.load(tmp_checkpoint_file.name))

    export_options = torch.onnx.ExportOptions(fake_context=fake_context)

    # `torch.onnx.dynamo_export` called WITHIN `torch.onnx.enable_fake_mode`
    export_output = torch.onnx.dynamo_export(
        fake_model,
        *fake_args,
        **fake_kwargs,
        export_options=export_options,
    )

    export_output.save("/path/to/model.onnx", model_state_dict=create_model())
```

If we decide to only support scenarios in which `torch._dynamo.export` is called within `FakeTensorMode`, then we can remove `fake_mode` argument from `torch._dynamo.export` as a follow-up task

ps: This PR is mostly Edward's https://github.com/pytorch/pytorch/pull/105468 + unit tests after an offline discussion
ps: https://github.com/pytorch/pytorch/issues/105464 tracks pending tasks/limitations from this PR

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

File: test/onnx/test_fx_to_onnx.py [hunk 1]
```diff
@@ -339,37 +339,20 @@ def forward(self, x):
...
 
-        # Scenario 3: Real model and fake input WITHOUT fake_context
-        with pytest.raises(torch.onnx.OnnxExporterError):
-            export_options = ExportOptions(fake_context=None)
-            _ = torch.onnx.dynamo_export(
-                real_model, fake_x, export_options=export_options
-            )
-
-        # Scenario 4: Real model and real input WITH fake_context
+        # Scenario 3: Real model and real input WITH fake_context
```
Comment (justinchuby) [inline test/onnx/test_fx_to_onnx.py:342]: Split each scenario into a different test method so they can fail and report individually? The fake mode tests can be in a new test class and the test `Model` can be initialized in `setUp()` for reuse if desired.

File: torch/onnx/_internal/exporter.py [hunk 2]
```diff
@@ -275,10 +275,12 @@ def enable_fake_mode():
...
 
     # This overrides the internal `FakeTensorMode` instance created by `torch._dynamo.export`[1].
-    # Ideally we should keep them in sync to preserve the same default behavior
+    # It is a good idea to keep them in sync (constructor args) to maintain the same default behavior
     # [1] `torch/_dynamo/output_graph.py::InstructionTranslator::OutputGraph.__init__`
+    # Mixed fake/real tensors are only allowed when `torch.onnx.dynamo_export` is not called within `FakeTensorMode`
+    # This is needed because models can create new parameters during `forward(self, *args, **kwargs)` run
     fake_mode = fake_tensor.FakeTensorMode(
-        allow_non_fake_inputs=True,  # https://github.com/pytorch/pytorch/issues/105077
+        allow_non_fake_inputs=not torch._guards.detect_fake_mode(),
```
Comment (ezyang) [inline torch/onnx/_internal/exporter.py:283]: you shouldn't need this anymore

**Review by ezyang — APPROVED** [review]
Can you add some tests for this that don't go through ONNX export pass? Similar to the samples we were passing around in Slack. You can put them with the regular export tests. Tahnks.

File: torch/onnx/_internal/exporter.py (hunk 2 above)
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:283]: `allow_non_fake_inputs=True` still produces

```
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 1467, in validate
    raise Exception(
Exception: Please convert all Tensors to FakeTensors first or instantiate FakeTensorMode with 'allow_non_fake_inputs'. Found in aten.embedding.default(Parameter(FakeTensor(..., size=(1024, 2), requires_grad=True)), tensor([...], size=(1, 3)))` 
```
in some tests (e.g. `test/onnx/test_fx_to_onnx_with_onnxruntime.py::TestFxToOnnxFakeTensorWithOnnxRuntime_op_level_debug_False_dynamic_shapes_False_load_checkpoint_during_init_True_export_within_fake_mode_False::test_large_scale_exporter_with_tiny_gpt2`)

File: test/onnx/test_fx_to_onnx.py (hunk 1 above)
Comment (thiagocrepaldi) [reply test/onnx/test_fx_to_onnx.py:342]: Added a TODO for future rework

File: torch/onnx/_internal/exporter.py (hunk 2 above)
Comment (ezyang) [reply torch/onnx/_internal/exporter.py:283]: 🤔 

Are you sure the exported model actually works if you run it under a fake mode with your inputs

Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:283]: Yes, we make numerical validation between ONNX Runtime and PyTorch Eager run.

The failing tests are the ones in which the model/input load are done within fake tensor mode, but `torch._dynamo.export` is not called within the fake mode

Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:283]: That is a repro:

```
    def test_capture_symbolic_tracing_outside_fake_mode(self):
        from torch._dynamo.output_graph import config
        from torch._subclasses import fake_tensor
        from torch.fx.experimental.symbolic_shapes import ShapeEnv

        class Model(torch.nn.Module):
            def __init__(self) -> None:
                super().__init__()
                self.linear = torch.nn.Linear(2, 2)
                self.linear2 = torch.nn.Linear(2, 2)

            def forward(self, x):
                out = self.linear(x)
                out = self.linear2(out)
                return out

        # User-instantiated FakeTensorMode
        fake_mode = fake_tensor.FakeTensorMode(
            allow_non_fake_inputs=False,
            allow_fallback_kernels=True,
            shape_env=ShapeEnv(
                allow_scalar_outputs=config.capture_scalar_outputs,
                allow_dynamic_output_shape_ops=config.capture_dynamic_output_shape_ops,
                frame_id=0,
            ),
        )
        # Fakefy input+model before exporting it
        with fake_mode:
            x = torch.rand(5, 2, 2)
            model = Model()

        # Export the model with fake inputs and parameters
        for aten_graph in [True, False]:
            graph_module, _ = torch._dynamo.export(model, x, aten_graph=aten_graph)
            self.assertTrue(
                isinstance(graph_module, torch.fx.GraphModule),
                msg="test_capture_symbolic_tracing_within_fake_mode_aten_graph_"
                + str(aten_graph),
            )
```
which raises
```
___________________________________________________________ ExportTests.test_capture_symbolic_tracing_within_fake_mode ____________________________________________________________
Traceback (most recent call last):
  File "/opt/conda/envs/ptca/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/opt/conda/envs/ptca/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/opt/conda/envs/ptca/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/opt/pytorch/torch/testing/_internal/common_utils.py", line 2356, in wrapper
    method(*args, **kwargs)
  File "/opt/pytorch/test/dynamo/test_export.py", line 3100, in test_capture_symbolic_tracing_within_fake_mode
    graph_module, _ = torch._dynamo.export(model, x, aten_graph=aten_graph)
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 1044, in export
    graph = make_fx(
  File "/opt/pytorch/torch/fx/experimental/proxy_tensor.py", line 809, in wrapped
    t = dispatch_trace(wrap_key(func, args, fx_tracer, pre_dispatch), tracer=fx_tracer, concrete_args=tuple(phs))
  File "/opt/pytorch/torch/_compile.py", line 24, in inner
    return torch._dynamo.disable(fn, recursive)(*args, **kwargs)
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 307, in _fn
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/_dynamo/external_utils.py", line 17, in inner
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/fx/experimental/proxy_tensor.py", line 468, in dispatch_trace
    graph = tracer.trace(root, concrete_args)
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 307, in _fn
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/_dynamo/external_utils.py", line 17, in inner
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/fx/_symbolic_trace.py", line 817, in trace
    (self.create_arg(fn(*args)),),
  File "/opt/pytorch/torch/fx/experimental/proxy_tensor.py", line 485, in wrapped
    out = f(*tensors)
  File "<string>", line 1, in <lambda>
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 1040, in graph_with_interpreter
    return torch.fx.Interpreter(graph).run(*args)
  File "/opt/pytorch/torch/fx/interpreter.py", line 138, in run
    self.env[node] = self.run_node(node)
  File "/opt/pytorch/torch/fx/interpreter.py", line 195, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/opt/pytorch/torch/fx/interpreter.py", line 312, in call_module
    return submod(*args, **kwargs)
  File "/opt/pytorch/torch/fx/_symbolic_trace.py", line 795, in module_call_wrapper
    return self.call_module(mod, forward, args, kwargs)
  File "/opt/pytorch/torch/fx/experimental/proxy_tensor.py", line 425, in call_module
    return forward(*args, **kwargs)
  File "/opt/pytorch/torch/fx/_symbolic_trace.py", line 788, in forward
    return _orig_module_call(mod, *args, **kwargs)
  File "/opt/pytorch/torch/nn/modules/module.py", line 1522, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/pytorch/torch/nn/modules/module.py", line 1531, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/pytorch/torch/nn/modules/linear.py", line 114, in forward
    return F.linear(input, self.weight, self.bias)
  File "/opt/pytorch/torch/utils/_stats.py", line 20, in wrapper
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 1054, in __torch_dispatch__
    return func(*args, **kwargs)
  File "/opt/pytorch/torch/_ops.py", line 437, in __call__
    return self._op(*args, **kwargs or {})
  File "/opt/pytorch/torch/utils/_stats.py", line 20, in wrapper
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 1174, in __torch_dispatch__
    return self.dispatch(func, types, args, kwargs)
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 1285, in dispatch
    ) = self.validate_and_convert_non_fake_tensors(func, converter, args, kwargs)
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 1479, in validate_and_convert_non_fake_tensors
    args, kwargs = tree_map_only(
  File "/opt/pytorch/torch/utils/_pytree.py", line 397, in tree_map_only
    return tree_map(map_only(ty)(fn), pytree)
  File "/opt/pytorch/torch/utils/_pytree.py", line 327, in tree_map
    return tree_unflatten([fn(i) for i in flat_args], spec)
  File "/opt/pytorch/torch/utils/_pytree.py", line 327, in <listcomp>
    return tree_unflatten([fn(i) for i in flat_args], spec)
  File "/opt/pytorch/torch/utils/_pytree.py", line 378, in inner
    return f(x)
  File "/opt/pytorch/torch/_subclasses/fake_tensor.py", line 1474, in validate
    assert x.fake_mode is self, "Mixing fake modes NYI"
AssertionError: Mixing fake modes NYI

While executing %l__self___linear : [num_users=1] = call_module[target=L__self___linear](args = (%l_x_,), kwargs = {})
Original traceback:
  File "/opt/pytorch/test/dynamo/test_export.py", line 3079, in forward
    out = self.linear(x)


To execute this test, run the following from the base repo dir:
     python test/dynamo/test_export.py -k test_capture_symbolic_tracing_within_fake_mode

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```

Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:283]: Reverting https://github.com/pytorch/pytorch/pull/105477/files#diff-c663df72d7a648204ff256e604fe3c9b393e7e2d58204e3bc818d90a99712911L937 fixes the last repro above (`test_capture_symbolic_tracing_outside_fake_mode`) with `allow_non_fake_inputs=False`, but `test/onnx/test_fx_to_onnx_with_onnxruntime.py::TestFxToOnnxFakeTensorWithOnnxRuntime_op_level_debug_False_dynamic_shapes_False_load_checkpoint_during_init_True_export_within_fake_mode_False::test_large_scale_exporter_with_tiny_gpt2` still fails

File: torch/_dynamo/eval_frame.py [hunk 3]
```diff
@@ -851,7 +852,7 @@ def export(
 
         fake_mode (fake_tensor.FakeTensorMode): Use this fake_mode instead of creating an internal one.
         Useful during symbolic tracing, when user input is already fakefied. Implies free fake tensors
-        are allowed on `make_fx`.
+        are allowed on `make_fx`. `fake_mode` must contain a valid (not None) `shape_env` instance.
```
Comment (BowenBao) [inline torch/_dynamo/eval_frame.py:855]: > `fake_mode` must contain a valid (not None) `shape_env` instance.

any check to enforce this?

File: test/onnx/test_fx_to_onnx_with_onnxruntime.py [hunk 4]
```diff
@@ -824,6 +826,43 @@ def create_pytorch_only_extra_kwargs():
...
+
+
+@parameterized.parameterized_class(
+    **_parameterized_class_attrs_and_values_with_fake_options(),
+    class_name_func=_parameterize_class_name,
+)
+class TestFxToOnnxFakeTensorWithOnnxRuntime(onnx_test_common._TestONNXRuntime):
+    """ONNX export test for specific Fake Tensor scenarios
+
+    TODO: Should we merge this with  `TestFxToOnnxWithOnnxRuntime`? Considerably increases export time
```
Comment (BowenBao) [inline test/onnx/test_fx_to_onnx_with_onnxruntime.py:853]: is it because of two new flags thus 4x time?

Perhaps just add one scenario for now to cover the fake mode feature and that it plays nicely with others. This can be followup pr. IIRC the tests are pretty light and fast so it is not that bad.

File: torch/_dynamo/eval_frame.py [hunk 5]
```diff
@@ -462,7 +463,7 @@ def catch_errors(frame, cache_size, frame_state):
                     )
                     return hijacked_callback(frame, cache_size, hooks, frame_state)
 
-        with compile_lock:
+        with compile_lock, _disable_current_modes():
```
Comment (BowenBao) [inline torch/_dynamo/eval_frame.py:466]: For my learning why this change?

File: torch/_dynamo/eval_frame.py [hunk 6]
```diff
@@ -934,7 +936,7 @@ def dynamo_normalization_capturing_compiler(
         graph = gm
 
         nonlocal fake_mode, example_inputs
-        fake_mode = fake_mode or _guards.detect_fake_mode(inner_example_inputs)
+        fake_mode = fake_mode or _guards.detect_fake_mode()
```
Comment (BowenBao) [inline torch/_dynamo/eval_frame.py:939]: ditto, why removing this? is it because it is redundant to pass 'inner_example_inputs' now?

File: torch/_dynamo/variables/builder.py [hunk 7]
```diff
@@ -1086,6 +1086,7 @@ def wrap_unspecialized_primitive(self, value):
                     example_value = unspec_var.proxy.node.meta["example_value"]
                 if is_fake(example_value):
                     fake_tensor_value = example_value
+                    assert fake_tensor_value.fake_mode is self.tx.fake_mode
```
Comment (BowenBao) [inline torch/_dynamo/variables/builder.py:1089]: Should we add a more informative message?

File: torch/_dynamo/variables/builder.py [hunk 8]
```diff
@@ -1233,7 +1237,11 @@ def _clone_input(value):
         example_value = _clone_input(example_value)
         proxy.node.meta["example_value"] = example_value
         specialized_props = target_cls.specialize(example_value)
-        if isinstance(example_value, torch._subclasses.fake_tensor.FakeTensor):
+        # TODO: not sure about this fake mode test
```
Comment (BowenBao) [inline torch/_dynamo/variables/builder.py:1240]: I'm confused too. I thought mixed fake modes is not allowed? But it looks as if it is expected and no error raised if spotted.

**Review by BowenBao — COMMENTED** [review]
IIUC, this changes in core dynamo is to allow 
```python
with fake_mode:
    _dynamo.export(model, *args)
```
where as previously only below is allowed
```python
_dynamo.export(model, *args, fake_mode=fake_mode)
```

File: test/dynamo/test_export.py [hunk 9]
```diff
@@ -2996,6 +2996,32 @@ def forward(self, x):
     return pytree.tree_unflatten([slice_tensor, slice_tensor_3, slice_tensor_6, slice_tensor_9], self._out_spec)""",
         )
 
+    def test_capture_symbolic_tracing_simple_within_fake_mode(self):
```
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply test/dynamo/test_export.py:2999]: This test builds on the existing ` test_capture_symbolic_tracing` at https://github.com/pytorch/pytorch/pull/105477/files#diff-91cb5d56bb8b6759eb8540df226fe248c5f422c09edd0751555cf2dc626add33L2999

The `_simple` tips that this is a very simple model (onliner `x + torch.randn(3)`); `_within_fake_mode` tips that the symbolic tracing is captured within a `FakeTensorMode` context manager.

File: torch/_dynamo/eval_frame.py (hunk 3 above)
Comment (thiagocrepaldi) [reply torch/_dynamo/eval_frame.py:855]: good point, will add an assert

File: torch/onnx/_internal/exporter.py (hunk 2 above)
Comment (ezyang) [reply torch/onnx/_internal/exporter.py:283]: What happens if you have the fake mode enabled when you export?

File: torch/_dynamo/eval_frame.py (hunk 5 above)
Comment (thiagocrepaldi) [reply torch/_dynamo/eval_frame.py:466]: Maybe @ezyang can explain this, but IIRC this helped in finding the right fake mode instance during tracing

File: torch/_dynamo/variables/builder.py (hunk 8 above)
Comment (thiagocrepaldi) [reply torch/_dynamo/variables/builder.py:1240]: The original check would assume the fake tensor was created by the global fake mode, because dynamo was designed to instantiate a fake mode internally and use it, so it would never have a mismatch

Now, we allow fake modes to be called outside `torch._dynamo.export` too, so now this check was extended to make sure that if the user input is fake, the fake mode used to fakefy it matches the one configured for the `torch._dynamo.export` lifecycle.

This basically tightens the fake_mode checks within dynamo, it doesn't change any behavior

File: torch/_dynamo/eval_frame.py (hunk 5 above)
Comment (ezyang) [reply torch/_dynamo/eval_frame.py:466]: The basic concept is that when we are running the compiler stack, we don't actually want user modes to be turned on, they can interfere unpredictably with the stack. In this particular case, fake mode is causing the problem.

I am actually a bit surprised this passes all CI, because it is kind of risky, if we are depending on a different mode still being enabled internally for eg AOTAutograd...

File: torch/_dynamo/variables/builder.py (hunk 8 above)
Comment (BowenBao) [reply torch/_dynamo/variables/builder.py:1244]: Let's write it as check + raise when mixed fake mode is detected then?

Comment (thiagocrepaldi) [reply torch/_dynamo/variables/builder.py:1244]: That is already done early (before FX tracing begins) in the execution path within `FakeTensorMode.dispatch` call stack, which includes a `validate_and_convert_non_fake_tensors(func, converter, args, kwargs)` step.

Comment (thiagocrepaldi) [reply torch/_dynamo/variables/builder.py:1244]: @ezyang what do you think? should we check whether mixed fake mode happens here or leave it as is (just checking fake mode matches)?

**Review by BowenBao — APPROVED** [review]

File: torch/_dynamo/variables/builder.py (hunk 8 above)
Comment (ezyang) [reply torch/_dynamo/variables/builder.py:1255]: So, with the modifications from this PR, we have two modes but they should not mix.

The outer mode is the user defined fake tensor mode. Inside Dynamo, these should show up as ambient tensors that we want to fakeify. This mode typically does not have symbolic shapes.

The inner mode is Dynamo's administrative fake tensor mode. All tensors get converted into this mode before we do reasoning on it. This mode should have symbolic shapes if you're modeling dynamism.

`wrap_fx_proxy_cls` can see both types of fake tensor. It gets called to wrap preexisting tensors (outer mode). It also gets called recursively after Dynamo has already fakeified a result (inner mode).
